libraries/AP_HAL_Linux/UARTDriver.cpp

```diff
@@ -284,44 +244,33 @@ void UARTDriver::set_blocking_writes(bool blocking)
  */
 bool UARTDriver::tx_pending()
 {
-    return !BUF_EMPTY(_writebuf);
+    return (_writebuf.available() > 0);
 }
 
 /*
   return the number of bytes available to be read
  */
 uint32_t UARTDriver::available()
 {
-    if (!_initialised) {
-        return 0;
-    }
-    uint16_t _tail;
-    return BUF_AVAILABLE(_readbuf);
+    return _readbuf.available();
 }
 
 /*
   how many bytes are available in the output buffer?
  */
 uint32_t UARTDriver::txspace()
 {
-    if (!_initialised) {
-        return 0;
-    }
-    uint16_t _head;
-    return BUF_SPACE(_writebuf);
+    return _writebuf.space();
 }
 
 int16_t UARTDriver::read()
 {
     uint8_t c;
-    if (!_initialised || _readbuf == NULL) {
-        return -1;
-    }
-    if (BUF_EMPTY(_readbuf)) {
+    if (!_initialised) {
         return -1;
     }
-    c = _readbuf[_readbuf_head];
-    BUF_ADVANCEHEAD(_readbuf, 1);
+    c = _readbuf.peek(0);
+    _readbuf.advance(1);
```

Comment:
I thought that there was comments about the read case as well, anyhow. For a single byte, read will do a lot of operations to handle more complex case of buffer splits and so on. Since we have peek for other performance reasons, original code did a clever use of those methods to perform better on single byte reads. This case really was nothing to do with the _write_fd one. That's why advance is called right away. But in both cases, read() would bring perfomance penalts that aren't desired in this part of the code even if it "looks" better.